textattack/augmentation/augmenter.py

```diff
@@ -11,40 +15,100 @@ class Augmenter:
             that suggests new texts from an input.
         constraints: (list(textattack.Constraint)): constraints
             that each transformation must meet
+        words_to_swap: (int): Number of words to swap per augmented example
+        transformations_per_example: (int): Maximum number of augmentations
+            per input
     """
-    def __init__(self, transformation, constraints=[]):
+    def __init__(self, transformation, constraints=[], words_to_swap=1, 
+        transformations_per_example=1):
         self.transformation = transformation
-        self.constraints = constraints
+        self.words_to_swap = words_to_swap
+        self.transformations_per_example = transformations_per_example
+        
+        self.constraints = []
+        self.pre_transformation_constraints = []
+        for constraint in constraints:
+            if isinstance(constraint, PreTransformationConstraint):
+                self.pre_transformation_constraints.append(constraint)
+            else:
+                self.constraints.append(constraint)
     
     def _filter_transformations(self, tokenized_text, transformations):
-        """ 
-        Filters a list of `TokenizedText` objects to include only the ones 
-        that pass `self.constraints`.
+        """ Filters a list of `TokenizedText` objects to include only the ones 
+            that pass `self.constraints`.
         """
         for C in self.constraints:
             if len(transformations) == 0: break
             transformations = C.call_many(tokenized_text, transformations, original_text=tokenized_text)
         return transformations
     
     def augment(self, text):
-        """ 
-        Returns all possible augmentations of `text` according to 
-        `self.transformation`.
+        """ Returns all possible augmentations of `text` according to 
+            `self.transformation`.
         """
         tokenized_text = TokenizedText(text, DummyTokenizer())
-        # Get potential transformations for text.
-        transformations = self.transformation(tokenized_text)
-        # Filter out transformations that don't match the constraints.
-        transformations = self._filter_transformations(tokenized_text, transformations)
-        return [t.clean_text() for t in transformations]
+        all_transformations = set()
+        for _ in range(self.transformations_per_example):
+            indices_to_modify = set(range(len(tokenized_text.words)))
+            next_tokenized_text = tokenized_text
+            for __ in range(self.words_to_swap):
+                transformations = []
+                for constraint in self.pre_transformation_constraints:
```

Comment:
Hmm I see why you need to apply these before calling the transformation the way you have it set up. This way works for everything but `CompositeTransformation`s. Is there a problem with just calling the transformation `self.words_to_swap` times and choosing a random one each time? That way you could use the current method we have for applying pre_modification_constraints and the same index won't be replaced twice since you're enforcing the RepeatModification constraint.

Reply:
I guess this would be a bit slower ugh. Another option is to keep what you have now but just try all indices, and if you find an index that gives you 0 transformation options keep choosing random ones until you find one that doesn't. While keeping the break if there are no indices left.